Perform Alignment:
	Usage: sample_verification.sh <RUN_ID> (e.g. sample_verification.sh RUN001)
	
	File required from DB: sample_list_unverified_<RUN_ID>.txt (e.g. sample_list_unverified_RUN001.txt)
	File location: /n/groups/cbdm-db/rnaseq_db/Sample_List_Info/
	File Format: Sample_Name<Tab>Sample_ID<Tab>Owner_ID<Tab>Species (e.g. B.Sp.Listeria.ip.Day0#1	000008508569_C01	Liang	Musmusculus)
	
	Outputs:
		/n/groups/cbdm-db/rnaseq_db/Sample_List_Info/sample_list_<RUN_ID>_<Species>.txt (Samples with available FASTQs)
		/n/groups/cbdm-db/rnaseq_db/Sample_List_Info/sample_list_report_<RUN_ID>.txt (Samples without available FASTQs)
		
		/n/groups/cbdm-db/rnaseq_db/output/alignment/<RUN_ID>/<Species>/BAMs/<Sample_ID>/<Sample_ID>.sortedByCoord.bam
		/n/groups/cbdm-db/rnaseq_db/output/alignment/<RUN_ID>/<Species>/BAMs/<Sample_ID>/<Sample_ID>.sortedByCoord.bam.bai
		/n/groups/cbdm-db/rnaseq_db/output/alignment/<RUN_ID>/<Species>/Count_Tables/<Sample_ID>/<Sample_ID>.featureCounts.txt
		/n/groups/cbdm-db/rnaseq_db/output/alignment/<RUN_ID>/<Species>/QC_data/multiqc_star.txt
		/n/groups/cbdm-db/rnaseq_db/output/alignment/<RUN_ID>/<Species>/QC_data/multiqc_featureCounts.txt
		
Perform Normalization:
	Usage: normalization.sh <RUN_ID> (e.g. normalization.sh RUN001)
	
	File required: sample_list_<RUN_ID>_<Species>.txt (e.g. sample_list_RUN001_Musmusculus.txt generated by sample_verification.sh)
	File location: /n/groups/cbdm-db/rnaseq_db/Sample_List_Info/
	File Format: Sample_Name<Tab>Sample_ID<Tab>Owner_ID<Tab>Species (e.g. B.Sp.Listeria.ip.Day0#1	000008508569_C01	Liang	Musmusculus)
	
	Outputs:
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/genes_count_table.tsv (raw gene count table)
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/genes_count_table_QC_passed.tsv (gene count table only including QC-passed samples)
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/sample_list.txt
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/normalized_gene_expression_1_adjust.gct
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/normalized_gene_expression_1_adjust.cls
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/Reads_Mapping_Quantification_Statistics.xlsx
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/RNASeq_QC_Report.html
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/gene_expression_data.csv
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/sample_expression_data.csv
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/Pearson_Correlation_Matrix_Samples.csv
		
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/genes_count_table.tsv (raw gene count table)
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/genes_count_table_QC_passed.tsv (gene count table only including QC-passed samples)
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/sample_list.txt
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/normalized_gene_expression_1_adjust.gct
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/normalized_gene_expression_1_adjust.cls
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/Reads_Mapping_Quantification_Statistics.xlsx
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/RNASeq_QC_Report.html
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/gene_expression_data.csv
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/sample_expression_data.csv
		/n/groups/cbdm-db/rnaseq_db/output/normalization/<RUN_ID>/<Species>/<Owner_ID>/Pearson_Correlation_Matrix_Samples.csv
		
Analyze Datasets:
	Usage: analyze_datasets.sh <USER_ID> <TIMESTAMP> (e.g. analyze_datasets.sh Liang 201803071032)
	
	File required: sample_list_<USER_ID>_<TIMESTAMP>.txt (e.g. sample_list_Liang_201803071032.txt)
	File location: /n/groups/cbdm-db/rnaseq_db/Sample_List_Info/
	File Format: Sample_Name<Tab>Sample_ID<Tab>RUN_ID<Tab>Species (e.g. B.Sp.Listeria.ip.Day0#1	000008508569_C01	RUN001	Musmusculus)
	
	Outputs:
		/n/groups/cbdm-db/rnaseq_db/output/analyze_datasets/<USER_ID>/genes_count_table_<USER_ID>_<TIMESTAMP>.tsv
		/n/groups/cbdm-db/rnaseq_db/output/analyze_datasets/<USER_ID>/sample_list_<USER_ID>_<TIMESTAMP>.txt
		/n/groups/cbdm-db/rnaseq_db/output/analyze_datasets/<USER_ID>/normalized_gene_expression_1_adjust_<USER_ID>_<TIMESTAMP>.gct
		/n/groups/cbdm-db/rnaseq_db/output/analyze_datasets/<USER_ID>/normalized_gene_expression_1_adjust_<USER_ID>_<TIMESTAMP>.cls
		
