---
title: "RNA-Seq QC Report"
author: "CBDM Lab"
output: 
    html_document:
        toc: TRUE
params: 
  owner_id: "NAMESPACE"
  owner_status: "NAMESPACE"
  RUN_ID: "NAMESPACE"
  SPECIES: "NAMESPACE"
---

	```{r include = FALSE}
	#####################################################################################################################

	### Get Parameters
	owner_id=params$owner_id
	owner_status=params$owner_status
	RUN_ID=params$RUN_ID
	SPECIES=params$SPECIES
	
	### Import MultiQC Data
	multiqc_star_all <- read.table(file=paste("/n/groups/cbdm-db/rnaseq_db/output_tmp/normalization/",RUN_ID,"/",SPECIES,"/multiqc_star.txt",sep=""),header=TRUE, comment.char = "", check.names = F, as.is = T)
	multiqc_featureCounts_all <- read.table(file=paste("/n/groups/cbdm-db/rnaseq_db/output_tmp/normalization/",RUN_ID,"/",SPECIES,"/multiqc_featureCounts.txt",sep=""),header=TRUE, comment.char = "", check.names = F, as.is = T)
	if(owner_status) {
		sample_info_all <- read.table(file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/sample_list.txt",sep=""), header=FALSE, stringsAsFactors = FALSE, comment.char = "", check.names = F, as.is = T)
	} else {
		sample_info_all <- read.table(file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/sample_list.txt",sep=""), header=FALSE, stringsAsFactors = FALSE, comment.char = "", check.names = F, as.is = T)
	}
	colnames(sample_info_all) <- c("Sample_Name", "Sample", "Owner")
	rownames(sample_info_all) <- sample_info_all$Sample_Name
	
	### Import Gene Counts Table
	if(owner_status) {
		out_tbl <- read.table(file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/genes_count_table.tsv",sep=""),header=TRUE, comment.char = "", check.names = F, as.is = T)
	} else {
		out_tbl <- read.table(file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/genes_count_table.tsv",sep=""),header=TRUE, comment.char = "", check.names = F, as.is = T)
	}
	sample_names <-  data.frame(Sample_Name=colnames(out_tbl[,-1]), stringsAsFactors = FALSE)
	
	### Extract Metadata and QC data
	sample_info <- merge(sample_names, sample_info_all)
	multiqc_star <- merge(sample_info, multiqc_star_all)
	multiqc_featureCounts <- merge(sample_info, multiqc_featureCounts_all)
	
	multiqc_star_output <- multiqc_star[, c("Sample_Name","Sample","total_reads","uniquely_mapped","uniquely_mapped_percent")]
	multiqc_featureCounts_output <- multiqc_featureCounts[, c("Sample_Name", "Sample", "Assigned", "percent_assigned", "Unassigned_Ambiguity", "Unassigned_MultiMapping", "Unassigned_NoFeatures")]
	
	multiqc_star_barchart_output <- multiqc_star[, c("Sample_Name", "uniquely_mapped_percent", "multimapped_percent", "multimapped_toomany_percent", "unmapped_tooshort_percent", "unmapped_other_percent")]
	multiqc_star_barchart_output_merged <- as.data.frame(tbl_df(multiqc_star_barchart_output) %>% mutate(Multi_mapped_rate = multimapped_percent + multimapped_toomany_percent) %>% mutate(Unmapped_rate = unmapped_tooshort_percent + unmapped_other_percent) %>% select(Sample_Name, Uniquely_mapped_rate = uniquely_mapped_percent, Multi_mapped_rate, Unmapped_rate))
	multiqc_featureCounts_barchart_output <- multiqc_featureCounts[, c("Sample_Name", "Assigned", "Unassigned_Ambiguity", "Unassigned_MultiMapping", "Unassigned_NoFeatures", "Total")]
	multiqc_featureCounts_barchart_output_merged <- as.data.frame(tbl_df(multiqc_featureCounts_barchart_output) %>% mutate(Assigned_rate = Assigned/Total*100) %>% mutate(Unassigned_Ambiguity_rate = Unassigned_Ambiguity/Total*100) %>% mutate(Unassigned_MultiMapping_rate = Unassigned_MultiMapping/Total*100) %>% mutate(Unassigned_NoFeatures_rate = Unassigned_NoFeatures/Total*100) %>% select(Sample_Name, Assigned_rate, Unassigned_Ambiguity_rate, Unassigned_MultiMapping_rate, Unassigned_NoFeatures_rate))
	
	#####################################################################################################################
	
	### QC Filtering (Uniquely mapped reads > 1 million)
	sample_names_passing <- as.vector(multiqc_star_output$Sample_Name[multiqc_star_output$uniquely_mapped>1000000])
	sample_names_all <- as.vector(multiqc_star_output$Sample_Name)
	
	if(length(sample_names_passing) == 1){
		counts <- as.data.frame(out_tbl[, sample_names_passing])
		colnames(counts) = sample_names_passing
		counts_output <- data.frame(out_tbl[,1], counts, stringsAsFactors = FALSE)
		colnames(counts_output) = c("Gene_Symbol", sample_names_passing)
	} else {
		counts <- out_tbl[, sample_names_passing]
		counts_output <- data.frame(out_tbl[,1], counts, stringsAsFactors = FALSE)
		colnames(counts_output) = c("Gene_Symbol", colnames(counts))
	}
	
	counts_all <- out_tbl[, sample_names_all]	
	
	### Output Gene Counts Table with QC-passed Samples
	if(owner_status) {
		write.table(counts_output, file = paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/genes_count_table_QC_passed.tsv",sep=""), quote=FALSE, row.names=FALSE, sep="\t")
	} else {
		write.table(counts_output, file = paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/genes_count_table_QC_passed.tsv",sep=""), quote=FALSE, row.names=FALSE, sep="\t")
	}
	
	gene_names <- out_tbl[,1]
	rownames(counts) <- gene_names
	rownames(counts_all) <- gene_names
	
	num_of_genes_at_least_1_hit <- as.data.frame(sapply(out_tbl[,sample_names_all], function(x)sum(x>0)))
	num_of_genes_over_10_hits <- as.data.frame(sapply(out_tbl[,sample_names_all], function(x)sum(x>10)))
	num_of_genes_over_50_hits <- as.data.frame(sapply(out_tbl[,sample_names_all], function(x)sum(x>50)))
	num_of_genes_mapped <- cbind(sample_names_all, num_of_genes_at_least_1_hit, num_of_genes_over_10_hits, num_of_genes_over_50_hits)
	colnames(num_of_genes_mapped) <- c("Sample_Name", "Number of genes with at least 1 mapped reads", "Number of genes with over 10 mapped reads", "Number of genes with over 50 mapped reads")
		
	#####################################################################################################################
	
	### Generate CLS file
	Num_of_samples <- length(sample_names_passing)
	sample_of_class <- unique(unlist(lapply(sample_names_passing, function(x) strsplit(x, split="#")[[1]][1])))
	Num_of_class <- length(sample_of_class)
	sample_index <- matrix(match(unlist(lapply(sample_names_passing, function(x) strsplit(x, split="#")[[1]][1])), sample_of_class)-1,nrow=1,byrow=TRUE)
	cls_header_1 <- matrix(c(Num_of_samples, Num_of_class, 1),nrow=1,byrow=TRUE)
	cls_header_2 <- matrix(c('#', sample_of_class),nrow=1,byrow=TRUE)
	if(owner_status) {
		write.table(cls_header_1, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/normalized_gene_expression_1_adjust.cls",sep=""),quote=FALSE,sep = ' ',row.names=FALSE,col.names=FALSE)
		write.table(cls_header_2, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/normalized_gene_expression_1_adjust.cls",sep=""),quote=FALSE,sep = ' ',row.names=FALSE,col.names=FALSE,append=TRUE)
		write.table(sample_index, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/normalized_gene_expression_1_adjust.cls",sep=""),sep=' ',quote=FALSE, row.names=FALSE,col.names=FALSE,append=TRUE)
	} else {
		write.table(cls_header_1, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/normalized_gene_expression_1_adjust.cls",sep=""),quote=FALSE,sep = ' ',row.names=FALSE,col.names=FALSE)
		write.table(cls_header_2, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/normalized_gene_expression_1_adjust.cls",sep=""),quote=FALSE,sep = ' ',row.names=FALSE,col.names=FALSE,append=TRUE)
		write.table(sample_index, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/normalized_gene_expression_1_adjust.cls",sep=""),sep=' ',quote=FALSE, row.names=FALSE,col.names=FALSE,append=TRUE)
	}
	
	### Normalization by DESeq2 and Generate GCT file
	pData <- sample_info_all[sample_names_passing,]
	deseqobj <- DESeqDataSetFromMatrix(countData=counts, colData=pData, design= ~ 1)
	dds <- estimateSizeFactors(deseqobj)
	normalized.counts <-counts(dds, normalized=TRUE)
	
	dat <- normalized.counts
	dat <- dat+1
	pt1_adjust <- data.frame(gene_names,gene_names,dat)
	col_names <- c('NAME','DESCRIPTION',sample_names_passing)
	names(pt1_adjust) <- col_names
	gct_header_1 <- matrix(c('#1.2'),nrow=1,byrow=TRUE)
	gct_header_2 <- matrix(c(dim(dat)[1],dim(dat)[2]),nrow=1,byrow=TRUE)
	if(owner_status) {
		write.table(gct_header_1, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/normalized_gene_expression_1_adjust.gct",sep=""),quote=FALSE,sep = '\t',row.names=FALSE,col.names=FALSE)
		write.table(gct_header_2, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/normalized_gene_expression_1_adjust.gct",sep=""),quote=FALSE,sep = '\t',row.names=FALSE,col.names=FALSE,append=TRUE)
		write.table(pt1_adjust,file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/normalized_gene_expression_1_adjust.gct",sep=""),sep='\t',quote=FALSE,row.names=FALSE,append=TRUE)
	} else {
		write.table(gct_header_1, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/normalized_gene_expression_1_adjust.gct",sep=""),quote=FALSE,sep = '\t',row.names=FALSE,col.names=FALSE)
		write.table(gct_header_2, file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/normalized_gene_expression_1_adjust.gct",sep=""),quote=FALSE,sep = '\t',row.names=FALSE,col.names=FALSE,append=TRUE)
		write.table(pt1_adjust,file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/normalized_gene_expression_1_adjust.gct",sep=""),sep='\t',quote=FALSE,row.names=FALSE,append=TRUE)
	}
	
	#####################################################################################################################
	
	sample_of_class_all <- unique(unlist(lapply(sample_names_all, function(x) strsplit(x, split="#")[[1]][1])))
	sample_index_all <- matrix(match(unlist(lapply(sample_names_all, function(x) strsplit(x, split="#")[[1]][1])), sample_of_class_all)-1,nrow=1,byrow=TRUE)

	pData_all <- sample_info_all[sample_names_all,]
	deseqobj_all <- DESeqDataSetFromMatrix(countData=counts_all, colData=pData_all, design= ~ 1)
	dds_all <- estimateSizeFactors(deseqobj_all)
	normalized.counts_all <-counts(dds_all, normalized=TRUE)
	
	### calculate mean, median and dynamic of expression across genes
	gene_means <- format(rowMeans(normalized.counts_all), digits = 2, nsmall = 2, scientific = F, trim = T)
	gene_medians <- format(rowMedians(normalized.counts_all), digits = 2, nsmall = 2, scientific = F, trim = T)
	gene_dynamic_ranges <- format(apply(normalized.counts_all+1, 1, function(x)max(x)/min(x)), digits = 2, nsmall = 2, scientific = F, trim = T)
	gene_expression_data <- data.frame(gene_names,gene_means,gene_medians,gene_dynamic_ranges)
	names(gene_expression_data) <- c('gene_names','gene_means','gene_medians','gene_dynamic_ranges')
	
	###calculate mean, median and dynamic of expression across samples
	sample_means <- format(colMeans(normalized.counts_all), digits = 2, nsmall = 2, big.mark = ",", trim = T)
	sample_medians <- format(colMedians(normalized.counts_all), digits = 2, nsmall = 2, big.mark = ",", trim = T)
	sample_dynamic_ranges <- format(apply(normalized.counts_all+1, 2, function(x)max(x)/min(x)), digits = 2, nsmall = 2, big.mark = ",", trim = T)
	sample_data <- data.frame(names(normalized.counts_all[1,]),sample_means,sample_medians,sample_dynamic_ranges)
	names(sample_data) <- c('sample_names','sample_means','sample_medians','sample_dynamic_ranges')
	
	### calculate Pearson correlation matrix across samples
	corr_data <- cor(normalized.counts_all)
	corr_data_output <- data.frame(rownames(corr_data),corr_data)
	colnames(corr_data_output) = c("Sample_Names", colnames(corr_data))
	
	if(owner_status) {
		write.table(gene_expression_data,file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/gene_expression_data.csv",sep=""),row.names=FALSE,quote=FALSE,sep=',')
		write.table(sample_data,file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/sample_expression_data.csv",sep=""),quote=FALSE,row.names=FALSE,sep=',')
		write.table(corr_data_output, quote=FALSE, row.names=FALSE, file = paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/Pearson_Correlation_Matrix_Samples.csv",sep=""), sep=',')
	} else {
		write.table(gene_expression_data,file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/gene_expression_data.csv",sep=""),row.names=FALSE,quote=FALSE,sep=',')
		write.table(sample_data,file=paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/sample_expression_data.csv",sep=""),quote=FALSE,row.names=FALSE,sep=',')
		write.table(corr_data_output, quote=FALSE, row.names=FALSE, file = paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/Pearson_Correlation_Matrix_Samples.csv",sep=""), sep=',')
	}
	
	corr_data_tmp <- data.frame(rownames(corr_data),corr_data, as.vector(sample_index_all), stringsAsFactors = F)
	colnames(corr_data_tmp) = c("Sample_Names", colnames(corr_data), "Class")
	replicates_corr_mean <- c()
	for(i in 1:length(corr_data_tmp$Sample_Names)){
	    replicates_corr_all <- corr_data_tmp[i ,colnames(corr_data_tmp) %in% corr_data_tmp$Sample_Names[corr_data_tmp$Class == corr_data_tmp$Class[i]]]
	    if(length(replicates_corr_all) > 1){
	        replicates_corr_mean = c(replicates_corr_mean, (rowSums(replicates_corr_all)-1)/(length(replicates_corr_all)-1))
	    } else {
	        replicates_corr_mean = c(replicates_corr_mean, replicates_corr_all)
	    }
	}
	replicates_corr_mean_output = data.frame(corr_data_tmp$Sample_Names, replicates_corr_mean)
	colnames(replicates_corr_mean_output) <- c("Sample_Name", "Average Correlation Coeffients Between Replicates")
	
	if(owner_status) {
		write.xlsx2(merge(merge(merge(multiqc_star_output,multiqc_featureCounts_output),num_of_genes_mapped),replicates_corr_mean_output), file = paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/",owner_id,"/Reads_Mapping_Quantification_Statistics.xlsx",sep=""), sheetName = "Statistics", row.names = F)
	} else {
		write.xlsx2(merge(merge(merge(multiqc_star_output,multiqc_featureCounts_output),num_of_genes_mapped),replicates_corr_mean_output), file = paste("/n/groups/cbdm-db/rnaseq_db/output/normalization/",RUN_ID,"/",SPECIES,"/Reads_Mapping_Quantification_Statistics.xlsx",sep=""), sheetName = "Statistics", row.names = F)
	}
	
	### 5-99 pearson correlation matrix
	gct_599 <- as.data.frame(normalized.counts_all)
	gct_599$rowmean <- rowMeans(gct_599)
	gct_599 <- gct_599[gct_599$rowmean>5 & gct_599$rowmean<quantile(gct_599$rowmean, 0.99), ]
	gct_599 <- select(gct_599, -rowmean)
	corr_data_599 <- cor(gct_599)

	```
	
## Reads Mapping Statistics

Statistics computed on raw sequencing reads, including total count, percent uniquely mapped, and how many reads did not map to the genome. Samples with less than 1 million uniquely mapped reads (highlighted by red color) are automatically excluded from normalization. Samples with 1~3 million uniquely mapped reads (highlighted by orange color) are included in the normalized output, but should be used with caution. 

```{r echo=FALSE}
datatable(multiqc_star_output) %>% formatCurrency(columns = c("total_reads","uniquely_mapped"), currency = "", interval = 3, digits = 0, mark = ",") %>% formatStyle(
  'uniquely_mapped',
  target = 'row',
  backgroundColor = styleInterval(c(1000000,3000000), c('red', 'orange', ''))
)

plot_ly(multiqc_star_barchart_output_merged, x = ~Sample_Name, y = ~Uniquely_mapped_rate, type = 'bar', name = 'Uniquely mapped rate') %>%
  add_trace(y = ~Multi_mapped_rate, name = 'Multi mapped rate') %>% add_trace(y = ~Unmapped_rate, name = 'Unmapped rate') %>%
  layout(yaxis = list(title = '% of reads'), xaxis = list(title = 'Sample Name'), barmode = 'stack', margin = list(b = 200))
```

## Mapped Gene Statistics

See table below for the number of genes with over one, ten, and fifty mapped reads. This is an indicator of genome coverage. For ImmGen samples, we recommend using only samples with over 8,000 genes over ten reads. Samples below this threshold should be used with caution.

```{r echo=FALSE}
datatable(format(num_of_genes_mapped, big.mark = ","))
```

## Quantification Statistics

Percent of reads assigned upon mapping, with resulting reason for unassigned reads.

```{r echo=FALSE}
datatable(multiqc_featureCounts_output) %>% formatCurrency(columns = c( "Assigned", "Unassigned_Ambiguity", "Unassigned_MultiMapping", "Unassigned_NoFeatures"), currency = "", interval = 3, digits = 0, mark = ",") %>% formatRound(columns = c("percent_assigned"))

plot_ly(multiqc_featureCounts_barchart_output_merged, x = ~Sample_Name, y = ~Assigned_rate, type = 'bar', name = 'Assigned rate') %>%
  add_trace(y = ~Unassigned_Ambiguity_rate, name = 'Unassigned Ambiguity rate') %>% add_trace(y = ~Unassigned_MultiMapping_rate, name = 'Unassigned MultiMapping rate') %>% add_trace(y = ~Unassigned_NoFeatures_rate, name = 'Unassigned NoFeatures rate') %>%
  layout(yaxis = list(title = '% of reads'), xaxis = list(title = 'Sample Name'), barmode = 'stack', margin = list(b = 200))
```

## Samples Expression Statistics

See below for the mean, median, and dynamic range of each sample across all genes. The dynamic range is a ratio of the largest to the smallest expression values for a sample.

```{r echo=FALSE}
datatable(sample_data, rownames= FALSE)
```

## Genes Expression Statistics

See below for the mean, median, and dynamic range of each gene across all samples. The dynamic range is a ratio of the largest to the smallest expression values for a gene.

```{r echo=FALSE, warning = FALSE}
datatable(gene_expression_data, rownames= FALSE)
```

## Sample Expression Correlation Heatmap

Heatmap of the Pearson correlations between all samples where the I,j_th entry provides the Pearson correlation between sample I and j. 

```{r echo=FALSE, warning = FALSE, message= FALSE}
heatmaply(corr_data, scale_fill_gradient_fun = ggplot2::scale_fill_gradientn(colors = c("#440154FF", "#21908CFF","#FDE725FF"), values = c(0, 0.79, 0.89, 1), guide = "colorbar", limits = c(0, 1), breaks = c(0, 0.8, 0.9, 1)), margins = c(150,200,40,20))
```

## Replicate Expression Correlation Heatmap

To avoid outlier effects, data were pre-processed by setting aside transcripts that had an average of fewer than five reads or were above the 99th percentile for number of reads in the dataset. The resulting heatmap of Pearson correlations between samples is shown below. For ImmGen samples, replicates should correlate at 0.9 or better.

```{r echo = FALSE, warning=FALSE, message=FALSE}
heatmaply(corr_data_599, scale_fill_gradient_fun = ggplot2::scale_fill_gradientn(colors = c("#440154FF", "#21908CFF","#FDE725FF"), values = c(0, 0.79, 0.89, 1), guide = "colorbar", limits = c(0, 1), breaks = c(0, 0.8, 0.9, 1)), margins = c(150,200,40,20))
```

## Marker Gene Heatmap

Heatmap showing normalized values of cell-type specific genes. Use as a basis to identify contamination in sorted samples.

```{r echo = FALSE, warning=FALSE, message=FALSE}
if(SPECIES=='Musmusculus'){
	contam_genes <- c("Xist", "Hba-a1", "Hbb-bs", "Jchain", "Ebf1", 
	                  "Prg2", "Pax5", "Tnfrsf17",
	                  "Cd3e", "Themis", "Lyz2", "Ighm", "Trdc", "Trac", "Fcgr1", "Msr1","Adam23",
	                  "Apol7c","Mc1r","Dkk3", "Chad", "Igha")
	contam_counts <- as.data.frame(normalized.counts_all)
	contam_counts$NAME <- rownames(contam_counts)
	contam <- contam_counts %>% filter(NAME %in% contam_genes)
	rownames(contam) <- contam$NAME
	contam <- dplyr::select(contam, -NAME)
	heatmaply(contam, margins = c(150,200,40,20), Rowv = F, Colv = F)
}

if(SPECIES=='Homosapiens'){
	contam_genes <- c("XIST", "HBA2", "HBB", "JCHAIN", "EBF1", 
	                  "PRG2", "PAX5", "TNFRSF17",
	                  "CD3E", "THEMIS", "LYZ", "FCGR1A", "MSR1","ADAM23",
	                  "MC1R","DKK3", "CHAD")
	contam_counts <- as.data.frame(normalized.counts_all)
	contam_counts$NAME <- rownames(contam_counts)
	contam <- contam_counts %>% filter(NAME %in% contam_genes)
	rownames(contam) <- contam$NAME
	contam <- dplyr::select(contam, -NAME)
	heatmaply(contam, margins = c(150,200,40,20), Rowv = F, Colv = F)
}
```
